node authorizer: limit kubelet access to ResourceClaim objects

pohly opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

As discussed during the KEP and code review of dynamic resource allocation for Kubernetes 1.26, to increase security kubelet should only get read access to those ResourceClaim objects that are needed by a pod on the node.

This can be enforced easily by the node authorizer:
- the names of the objects are defined by the pod status
- there is no indirection as with volumes (pod -> pvc -> pv)

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/113866

#### Does this PR introduce a user-facing change?
```release-note
kubelet: security of dynamic resource allocation was enhanced by limiting node access to those objects that are needed on the node.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
node authorizer: limit kubelet access to ResourceClaim objects

As discussed during the KEP and code review of dynamic resource allocation for
Kubernetes 1.26, to increase security kubelet should only get read access to
those ResourceClaim objects that are needed by a pod on the node.

This can be enforced easily by the node authorizer:
- the names of the objects are defined by the pod status
- there is no indirection as with volumes (pod -> pvc -> pv)

Normally the graph only gets updated when a pod is not scheduled yet.
Resource claim status changes can still happen after that, so they
also must trigger an update.

pohly commented:
Note that there is also [an issue](https://github.com/kubernetes/kubernetes/issues/113722) open to investigate whether the deterministic naming of generated ResourceClaim objects, something that this PR depends upon, can be replaced.

Suppose the name of the ResourceClaim for a pod was generated and "the right one" for a pod had a `pod.k8s.io=<pod name>` label (details TBD). Would the node authorizer still be able to to check access? Perhaps by adding graph edges when a ResourceClaim with such a label gets added?

But then how would kubelet learn about the right ResourceClaim? Can it watch "all" ResourceClaims with an informer and will it then only receive those that it has access to, if node authorizer checks that? I suppose that must work, because kubelet also needs to be informed efficiently about pods that were scheduled onto its node.

liggitt commented:
> But then how would kubelet learn about the right ResourceClaim? Can it watch "all" ResourceClaims with an informer and will it then only receive those that it has access to, if node authorizer checks that? I suppose that must work, because kubelet also needs to be informed efficiently about pods that were scheduled onto its node.

Pods are not narrowly authorized, actually... any node can list/watch all pods, they just happen to filter with a field selector to just the pods they care about. Field selectors are not in the set of information considered by authorization.

Pods are also a special case for scaling watch... they have a bespoke watch indexer hard-coded into the apiserver to make filtering by node name performant. I would not assume thousands of nodes watching all resource claims filtering by a label would perform well generically.

If the name can be deterministic, that would be much better both for the kubelet discovering and authorizing.

pohly commented:
> Pods are not narrowly authorized, actually...

Isn't that a pretty big gap in the security design? Limiting access to the other types is nice, but a compromised kubelet can still do a lot of damage when it can read/write any pod. But that's just an aside.

> I would not assume thousands of nodes watching all resource claims filtering by a label would perform well generically.

Suppose the node authorizer was modified to do authorization for ResourceClaim by label. Would a watch then a) send only authorized objects and b) would that be efficient?

liggitt commented:
> when it can read/write any pod

it can only write pods assigned to it, it can read any pod

> Suppose the node authorizer was modified to do authorization for ResourceClaim by label.

Authorizers don't have access to label selectors on incoming requests. Xref discussion in https://github.com/kubernetes/kubernetes/issues/44703#issuecomment-324826356

> Would a watch then a) send only authorized objects

A watch could send object that only matched a label selector, but authorization couldn't force a specific label selector

> and b) would that be efficient?

A cluster wide watch that sets a label selector that matches a fraction of the existing entries is not efficient. Every node starting a watch like this on an API that is high volume and fast moving (potentially 1:1 with pods) sounds problematic

pohly commented:
This all sounds like convincing arguments to keep the current design with predictable names.

It also means that this PR is ready for review and merging because we won't change the underlying mechanism.

liggitt commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -393,6 +396,16 @@ func (g *Graph) AddPod(pod *corev1.Pod) {
 	}
+
+	for _, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName := resourceclaim.Name(pod, &podResourceClaim)
```
it looks like resourceclaim.Name returns `pod.Name + claim.Name` ... since both of those can contain `-` characters, the following would conflict, right?

* pod `foo`, claim `bar-baz`
* pod `foo-bar`, claim `baz`

I also don't see validation that ensures any of the following:
* ResourceClaimName is valid (passes the same validation we apply in ValidateClaim)
* ResourceClaimTemplateName is valid (passes the same validation we apply in ValidateClaimTemplate)
* the combined `pod.Name + "-" + claim.Name` is valid (passes the same validation we apply in ValidateClaim, *especially* length requirements)

liggitt replied:
> The conflict between foo + bar-baz and foo-bar and baz was mentioned before. As both pods come from the same namespace, the user would be responsible for ensuring that no such clash occurs. Would it a) be possible to tighten the validation and b) useful?

I don't see why we would expose the user to sharp edges like that if we can avoid it with requirements for templated / directly used claim object names

pohly replied:
> I don't see why we would expose the user to sharp edges like that if we can avoid it with requirements for templated / directly used claim object names

So that's a "yes" for not allowing "-" in the `pod.spec.resourceClaims[*].name` field, right? Is it okay to tighten that validation in 1.27? I think it's okay as we are making some other "action required" changes, I just wanted to be sure.

> We should validate the fields we have. If we have a reference in core, we should validate it.

/cc @thockin 

I think you were the one who argued that core shouldn't validate against requirements defined by the v1alpha1 resource API.

I can add it back if the consensus is that we should do that.

klueska replied:
Can we instead use a separator that is disallowed in both? i.e. store the key as:
```
pod.Name + "#" + claim.Name
```

From a user's perspective it will feel unnecessarily prohibitive to not be allowed to use dashes in the claim name.

liggitt replied:
the issue isn't the key in the graph, it's the actual API object name

klueska replied:
Ah right.

pohly replied:
We can concatenate with `--` and then not allow that inside the claim name. The downside is that a generated `my-pod--my-claim` will look a bit odd. Or perhaps that's an upside? It further reduces the risk that someone accidentally creates a ResourceClaim with such a name and then has a conflict.

thockin replied:
I don't think we should lean on character-set trickery here.  I just went on a rant about this on Twitter - names are for humans.  Parsing names for metadata is a bad idea.

I'd much prefer that the name not be deterministic, and instead associate it by labels.

E.g.

Pod foo, claim bar-baz => "foo-random", + labels {"dra.k8s.io/pod: "foo", "dra.k8s.io/resource-claim: bar-baz"}
Pod foo-bar, claim baz => "foo-bar-random", + labels {"dra.k8s.io/pod: "foo-bar", "dra.k8s.io/resource-claim: baz"}

You need to handle the edge-case of getting more than one result, but that's solvable with docs.

Or instead of labels, generate a random name and store it in status?

pohly replied:
I was discussing such a solution earlier with Jordan (see above) and [the outcome was](https://github.com/kubernetes/kubernetes/pull/116254#issuecomment-1463202076):
> If the name can be deterministic, that would be much better both for the kubelet discovering and authorizing.

The reason why just labels doesn't work well is that a kubelet on a specific node won't know which ResourceClaim it needs to read for a pod - it'll have to list them with a label filter, which just won't be very efficient.

Storing the generated name in the status would work, but does anyone here really have the motivation for another core/v1 API change?

thockin replied:
> a kubelet on a specific node won't know which ResourceClaim it needs to read for a pod

OK, that's a fair point - ideally "The set of resources which need to be read in order to run a pod are directly linked by the pod".  Good principle.

> Storing the generated name in the status would work, but does anyone here really have the motivation for another core/v1 API change?

I do this all day.  v1 changes aren't SO bad given that you already have a gate in place.

@liggitt what think?

liggitt replied:
Either deterministic mappings or explicit references from pod status work in my view.

Putting the name of the spawned claim(s) in pod status for kubelets to discover and for the node authorizer to authorize on *could* work. `NodeRestriction` admission would need an update to disallow nodes adding names of claims in pod status, but that's not hard.

I don't have a good sense for the trust level we're placing in the controller(s) handling these claim objects and whether we'd be happy letting those controller(s) write to status of arbitrary pods. I haven't followed the design enough to know if we expect a single central controller running on the control plane (which we'd be happy to let write pod status), or if this is a distributed bring-your-own-controller model (where we would *not* want to hand out pod status update permissions like candy).

klueska replied:
It's a bring-your-own-controller model

liggitt replied:
sorry, I meant the controller creating the resourceclaim object and updating pod status

klueska replied:
Right -- that's launched by the `kube-controller-manager` and implemented here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/resourceclaim/controller.go

pohly commented:
I'd like to propose that we *not* merge this PR for 1.27. There's clearly a need for further discussion.

What may be useful now is adding the validation of the generated name and object references, without changing how things work.

furkatgofurov7 commented:
@pohly hi 👋🏼, v1.28 Bug Triage Lead here.

Reaching out to know if this PR is still on track for the 1.28 release? Although tide still thinks it must be in v1.27 milestone 🤔

pohly commented:
We need a decision about https://github.com/kubernetes/kubernetes/pull/117351 first and then potentially update this PR. I certainly hope that this is possible in 1.28.

furkatgofurov7 commented:
@pohly hi!

> We need a decision about #117351 first and then potentially update this PR. I certainly hope that this is possible in 1.28.

Great to hear that! Can we label this PR with a priority label? Also, just a reminder that the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

pohly commented:
/priority important-longterm

pohly commented:
I rebased after https://github.com/kubernetes/kubernetes/pull/117351 got merged today. This PR is ready for review and merging.

furkatgofurov7 commented:
@pohly 👋🏼

This is a reminder, as I informed about it already 2 weeks ago, we are in week 9 currently and the code freeze is starting next week (week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496).

This PR looks very close to the finish line, however can we make sure that this PR is on track and have sufficient labels in order to make it to release 1.28 before code freeze is in? Also, please let me know if you need any help/assistance.

liggitt commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -393,6 +396,20 @@ func (g *Graph) AddPod(pod *corev1.Pod) {
 	}
+
+	for _, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName, _, err := resourceclaim.Name(pod, &podResourceClaim)
```
it looks like this pulls from `.status.resourceClaimStatuses` ... does that get populated/updated before or after a pod is bound to a node? If so, this logic that assumes all the references in the pod are immutable after node binding needs to change:

https://github.com/kubernetes/kubernetes/blob/79c02ceb73f4d64f52a9d4c46785e4c7497493d9/plugin/pkg/auth/authorizer/node/graph_populator.go#L73-L90

pohly replied:
> does that get populated/updated before or after a pod is bound to a node?

Before. The scheduler will not schedule pods where claims are missing.

> all the references in the pod are immutable after node binding

They are, because the claims need to be allocated and reserved before scheduling, which implies that they have a finalizer and cannot disappear underneath the pod.

This is the normal flow. If a pod gets created with `spec.nodeName` already set, then kubelet may try to run containers before the claims are ready. In that case it will encounter various error conditions (claim not created, claim not allocated or not reserved, or permission to read it not granted yet) and retry until those problems are fixed. There is work under way to do this in the resource claim controller. Nontheless this obviously should better be avoided, but the authorizer changes should also be valid in such a scenario.

liggitt replied:
> the authorizer changes should also be valid in such a scenario

would they? in the current state of this PR, once the pod is processed with nodeName set, no updates to the pod (including status updates that would make generated resourceclaim objects resolveable) would be processed by the node authorizer (graphPopulator#updatePod short circuits if the update didn't change the pod UID or nodeName)

pohly replied:
> graphPopulator#updatePod short circuits if the update didn't change the pod UID or nodeName)

Hrm. That's a problem. Is extending that check to update also when `status.resourceClaimStatuses` changes okay?

pohly replied:
The E2E test in https://github.com/kubernetes/kubernetes/pull/118209 triggers this problem: kubelet keeps getting permission errors even after the status is updated.

liggitt replied:
> Is extending that check to update also when status.resourceClaimStatuses changes okay?

Doing another update when a meaningful change to resourceClaimStatuses happens is ok... ideally that check inside `updatePod` would be very cheap (since we want to ignore almost all pod updates to avoid churn in the authorizer graph). Would checking `len(pod.Status.ResourceClaimStatuses) == len(oldPod.Status.ResourceClaimStatuses)` or something similarly cheap catch the updates the controller would make?

pohly replied:
... and this change fixes it:

```patch
diff --git a/plugin/pkg/auth/authorizer/node/graph_populator.go b/plugin/pkg/auth/authorizer/node/graph_populator.go
index aff1f80c63c..2bed2ab84fa 100644
--- a/plugin/pkg/auth/authorizer/node/graph_populator.go
+++ b/plugin/pkg/auth/authorizer/node/graph_populator.go
@@ -23,6 +23,7 @@ import (
 
        corev1 "k8s.io/api/core/v1"
        storagev1 "k8s.io/api/storage/v1"
+       apiequality "k8s.io/apimachinery/pkg/api/equality"
        "k8s.io/apimachinery/pkg/util/wait"
        corev1informers "k8s.io/client-go/informers/core/v1"
        storageinformers "k8s.io/client-go/informers/storage/v1"
@@ -78,8 +79,9 @@ func (g *graphPopulator) updatePod(oldObj, obj interface{}) {
                return
        }
        if oldPod, ok := oldObj.(*corev1.Pod); ok && oldPod != nil {
-               if (pod.Spec.NodeName == oldPod.Spec.NodeName) && (pod.UID == oldPod.UID) {
-                       // Node and uid are unchanged, all object references in the pod spec are immutable
+               if (pod.Spec.NodeName == oldPod.Spec.NodeName) && (pod.UID == oldPod.UID) &&
+                       apiequality.Semantic.DeepEqual(oldPod.Status.ResourceClaimStatuses, pod.Status.ResourceClaimStatuses) {
+                       // Node and uid are unchanged, all object references in the pod spec are immutable respectively unmodified (claim statuses).
                        klog.V(5).Infof("updatePod %s/%s, node unchanged", pod.Namespace, pod.Name)
                        return
                }
```

I am not doing feature gate testing here, based on the recent discussion that if these feature-gated fields happen to be set, components consuming such a pod should support the feature.

`graphPopulator` doesn't seem to have unit tests. Do I need to add something or is it good enough to know that there will be E2E testing which covers this aspect?

pohly replied:
See https://github.com/kubernetes/kubernetes/compare/571a08aa844bbc8e0d0f44f1e2b87a72b0d4e510..53127ee125a5d42db0517983aad9d6b2efecc129

liggitt replied:
deep equality on a list of nested structs was what I was trying to avoid having to do on every update of every pod throughout the entire pod lifecycle

pohly commented:
While running E2E tests locally I noticed that there is a problem in kubelet after a pod has been force-deleted (removed from API server while running) : kubelet no longer has read-access to the resource claim of that pod, but tries to get it anyway and fails. IMHO it should do that GET, but I need to discuss that further with my fellow developers -> PR is WIP, will not make it for 1.28.

liggitt commented:
> kubelet no longer has read-access to the resource claim of that pod, but tries to get it anyway and fails

yeah, we don't want to block cleanup of removed pods on unavailability of related resources (those could fail for lots of reasons, being forbidden, being deleted, etc)

liggitt committed:
Add integration test for node authorizer claim references

pohly committed:
auth: don't allow kubelet to from modify ResourceClaimStatuses

The status determines which claims kubelet is allowed to access when claims get
created from a template. Therefore kubelet must not be allowed to modify that
part of the status, because otherwise it could add an entry and then gain
access to a claim it should have access to.

liggitt commented:
/lgtm
/approve

pohly commented:
@bart0sh: can you help getting this PR merged?

- get your kubelet checkpointing changes merged to get rid of the forbidden GET
- /test pull-kubernetes-kind-dra
- lift the hold when that is passing

/hold

bart0sh commented:
@pohly sure, here is my PR: https://github.com/kubernetes/kubernetes/pull/119307
I'll unhold this one when mine is merged.
Should it be rebased after mine is merged?

bart0sh commented:
/test pull-kubernetes-kind-dra

bart0sh commented:
/unhold
